### ISSUE TITLE
`duckdb.create_function` type inferencing doesn't work with `from __future__ import annotations`

### ISSUE THREAD

**Issue by earwig** [duckdb/duckdb#9226]
### What happens?

In the Python UDF interface, `create_function` is able to infer types by inspecting the signature of the function to be registered. However, this doesn't work with [PEP 563](https://peps.python.org/pep-0563/) string annotations. The reason is [we use `inspect.signature`](https://github.com/duckdb/duckdb/blob/0894f3d6ab1321619c40c5ef870ad252f16f83ab/tools/pythonpkg/src/python_udf.cpp#L282) which does not evaluate string annotations.

In Python 3.10+, [we can pass `eval_str=True`](https://docs.python.org/3/library/inspect.html#inspect.signature) to `inspect.signature` and it will parse the annotations for us. This has the disadvantage of potentially raising an exception or executing arbitrary code, but on balance, I think the user opts into this by registering a function without passing explicit types, and it is more useful than the current behavior in most cases.

### To Reproduce

```python
from __future__ import annotations
import duckdb

def foo(x: int) -> str:
    return str(x * 2)

duckdb.create_function("foo", foo)
```

This raises the following error:

```
Traceback (most recent call last):
  File "test.py", line 7, in <module>
    duckdb.create_function("foo", foo)
duckdb.duckdb.InvalidInputException: Invalid Input Error: Could not infer the return type, please set it explicitly
```

Removing the first line causes it to succeed.

### OS:

Linux

### DuckDB Version:

main

### DuckDB Client:

Python

### Full Name:

Ben Kurtovic

### Affiliation:

Unaffiliated

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by Tishj** [comment]
Just to clarify, your point is not about supporting the `__future__.annotations` module as a way to backport annotations to older python versions, but only about supporting the [PEP 563](https://peps.python.org/pep-0563/) aspect of this?

Because `eval_str` is only available since 3.10, so this wouldn't work in older versions anyways

**Comment by earwig** [comment]
Right. __future__.annotations isn’t used to backport annotations to older python versions, only to enable the string-style annotations, even in the latest Python version - as I understand it. I mentioned eval_str as a possible solution because it’s easy, though it limits the fix to 3.10. I think this is OK.On Oct 16, 2023, at 5:12 AM, Thijs ***@***.***> wrote:﻿
Just to clarify, your point is not about supporting the __future__.annotations module as a way to backport annotations to older python versions, but only about supporting the PEP 563 aspect of this?
Because eval_str is only available since 3.10, so this wouldn't work in older versions anyways
